File: carplay_dongle/readable.py
import struct
import json
import numpy as np
from enum import IntEnum
from typing import Optional, Dict, Any


from .common import MessageHeader, CommandMapping


class AudioCommand(IntEnum):
    AudioOutputStart = 1
    AudioOutputStop = 2
    AudioInputConfig = 3
    AudioPhonecallStart = 4
    AudioPhonecallStop = 5
    AudioNaviStart = 6
    AudioNaviStop = 7
    AudioSiriStart = 8
    AudioSiriStop = 9
    AudioMediaStart = 10
    AudioMediaStop = 11
    AudioAlertStart = 12
    AudioAlertStop = 13


class PhoneType(IntEnum):
    AndroidMirror = 1
    CarPlay = 3
    iPhoneMirror = 4
    AndroidAuto = 5
    HiCar = 6


class Message:
    def __init__(self, header: MessageHeader):
        self.header = header


class Command(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        command_value = struct.unpack('<I', data[0:4])[0]
        
        try:
            self.value = CommandMapping(command_value)
            print(f'📩 COMMAND: {self.value.name} ({command_value})')
        except ValueError:
            # Unknown command
            self.value = None
            print(f'⚠️  UNKNOWN COMMAND: {command_value} (0x{command_value:08x})')
            print(f'    Known commands range: 0-{max(CommandMapping).value}')
            
            # Show data if there's more
            if len(data) > 4:
                print(f'    Extra data length: {len(data) - 4} bytes')
                print(f'    Extra data (hex): {data[4:].hex()}')
            
            # Suggest possible command category
            if command_value < 100:
                print(f'    Category: Likely basic control command')
            elif 100 <= command_value < 200:
                print(f'    Category: Likely navigation/UI command')
            elif 200 <= command_value < 300:
                print(f'    Category: Likely media control command')
            elif 300 <= command_value < 400:
                print(f'    Category: Likely phone control command')
            elif command_value >= 1000:
                print(f'    Category: Likely wireless/connectivity command')
            
            # List similar known commands
            nearby_commands = [
                (name, val) for name, val in CommandMapping.__members__.items()
                if abs(val - command_value) <= 10
            ]
            if nearby_commands:
                print(f'    Nearby known commands:')
                for name, val in sorted(nearby_commands, key=lambda x: abs(x[1] - command_value)):
                    print(f'      - {name}: {val} (diff: {command_value - val:+d})')


class ManufacturerInfo(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.a = struct.unpack('<I', data[0:4])[0]
        self.b = struct.unpack('<I', data[4:8])[0]


class SoftwareVersion(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.version = data.decode('ascii')


class BluetoothAddress(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.address = data.decode('ascii')


class BluetoothPIN(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.pin = data.decode('ascii')


class BluetoothDeviceName(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.name = data.decode('ascii')


class WifiDeviceName(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.name = data.decode('ascii')


class HiCarLink(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.link = data.decode('ascii')


class BluetoothPairedList(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.data = data.decode('ascii')


class Plugged(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        wifi_avail = len(data) == 8
        
        if wifi_avail:
            self.phone_type = PhoneType(struct.unpack('<I', data[0:4])[0])
            self.wifi = struct.unpack('<I', data[4:8])[0]
            print(f'📱 PHONE PLUGGED: {self.phone_type.name}, WiFi available: {self.wifi}')
        else:
            self.phone_type = PhoneType(struct.unpack('<I', data[0:4])[0])
            self.wifi = None
            print(f'📱 PHONE PLUGGED: {self.phone_type.name}, WiFi not available')


class Unplugged(Message):
    def __init__(self, header: MessageHeader):
        super().__init__(header)
        print(f'📱 PHONE UNPLUGGED')


class AudioFormat:
    def __init__(self, frequency: int, channel: int, bit_depth: int, 
                 format_str: str = None, mime_type: str = None):
        self.frequency = frequency
        self.channel = channel
        self.bit_depth = bit_depth
        self.format = format_str
        self.mime_type = mime_type


DECODE_TYPE_MAP: Dict[int, AudioFormat] = {
    1: AudioFormat(44100, 2, 16, "S16LE", "audio/L16; rate=44100; channels=2"),
    2: AudioFormat(44100, 2, 16, "S16LE", "audio/L16; rate=44100; channels=2"),
    3: AudioFormat(8000, 1, 16, "S16LE", "audio/L16; rate=8000; channels=1"),
    4: AudioFormat(48000, 2, 16, "S16LE", "audio/L16; rate=48000; channels=2"),
    5: AudioFormat(16000, 1, 16, "S16LE", "audio/L16; rate=16000; channels=1"),
    6: AudioFormat(24000, 1, 16, "S16LE", "audio/L16; rate=24000; channels=1"),
    7: AudioFormat(16000, 2, 16, "S16LE", "audio/L16; rate=16000; channels=2"),
}


class AudioData(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.decode_type = struct.unpack('<I', data[0:4])[0]
        self.volume = struct.unpack('<f', data[4:8])[0]
        self.audio_type = struct.unpack('<I', data[8:12])[0]
        
        amount = len(data) - 12
        self.command = None
        self.volume_duration = None
        self.data = None
        
        if amount == 1:
            cmd_val = struct.unpack('<b', data[12:13])[0]
            try:
                self.command = AudioCommand(cmd_val)
            except ValueError:
                print(f'⚠️  UNKNOWN AUDIO COMMAND: {cmd_val}')
                print(f'    Known audio commands: {[c.name for c in AudioCommand]}')
        elif amount == 4:
            self.volume_duration = struct.unpack('<f', data[12:16])[0]
        else:
            # Convert to int16 array
            self.data = np.frombuffer(data[12:], dtype=np.int16)


class VideoData(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.width = struct.unpack('<I', data[0:4])[0]
        self.height = struct.unpack('<I', data[4:8])[0]
        self.flags = struct.unpack('<I', data[8:12])[0]
        self.length = struct.unpack('<I', data[12:16])[0]
        self.unknown = struct.unpack('<I', data[16:20])[0]
        self.data = data[20:]


class MediaType(IntEnum):
    Data = 1
    AlbumCover = 3


class MediaData(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        media_type = struct.unpack('<I', data[0:4])[0]
        
        self.payload = None
        
        if media_type == MediaType.AlbumCover:
            import base64
            image_data = data[4:]
            self.payload = {
                'type': MediaType.AlbumCover,
                'base64Image': base64.b64encode(image_data).decode('ascii')
            }
        elif media_type == MediaType.Data:
            media_data = data[4:-1]
            self.payload = {
                'type': MediaType.Data,
                'media': json.loads(media_data.decode('utf-8'))
            }
        else:
            print(f'⚠️  UNEXPECTED MEDIA TYPE: {media_type}')
            print(f'    Known media types: {[t.name for t in MediaType]}')


class Opened(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.width = struct.unpack('<I', data[0:4])[0]
        self.height = struct.unpack('<I', data[4:8])[0]
        self.fps = struct.unpack('<I', data[8:12])[0]
        self.format = struct.unpack('<I', data[12:16])[0]
        self.packet_max = struct.unpack('<I', data[16:20])[0]
        self.i_box = struct.unpack('<I', data[20:24])[0]
        self.phone_mode = struct.unpack('<I', data[24:28])[0]
        print(f'📺 SESSION OPENED: {self.width}x{self.height} @ {self.fps}fps')


class BoxInfo(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.settings = json.loads(data.decode('utf-8'))


class Phase(Message):
    def __init__(self, header: MessageHeader, data: bytes):
        super().__init__(header)
        self.phase = struct.unpack('<I', data[0:4])[0]
        print(f'🔄 PHASE CHANGE: {self.phase}')

File: src/carplay/messages/readable.js
// ============================================================================
// messages/readable.js
// ============================================================================
import { MessageHeader, CommandMapping } from './common.js'

export const AudioCommand = {
  AudioOutputStart: 1,
  AudioOutputStop: 2,
  AudioInputConfig: 3,
  AudioPhonecallStart: 4,
  AudioPhonecallStop: 5,
  AudioNaviStart: 6,
  AudioNaviStop: 7,
  AudioSiriStart: 8,
  AudioSiriStop: 9,
  AudioMediaStart: 10,
  AudioMediaStop: 11,
  AudioAlertStart: 12,
  AudioAlertStop: 13,
}

export class Message {
  constructor(header) {
    this.header = header
  }
}

export class Command extends Message {
  constructor(header, data) {
    super(header)
    this.value = data.readUInt32LE(0)
  }
}

export class ManufacturerInfo extends Message {
  constructor(header, data) {
    super(header)
    this.a = data.readUInt32LE(0)
    this.b = data.readUInt32LE(4)
  }
}

export class SoftwareVersion extends Message {
  constructor(header, data) {
    super(header)
    this.version = data.toString('ascii')
  }
}

export class BluetoothAddress extends Message {
  constructor(header, data) {
    super(header)
    this.address = data.toString('ascii')
  }
}

export class BluetoothPIN extends Message {
  constructor(header, data) {
    super(header)
    this.pin = data.toString('ascii')
  }
}

export class BluetoothDeviceName extends Message {
  constructor(header, data) {
    super(header)
    this.name = data.toString('ascii')
  }
}

export class WifiDeviceName extends Message {
  constructor(header, data) {
    super(header)
    this.name = data.toString('ascii')
  }
}

export class HiCarLink extends Message {
  constructor(header, data) {
    super(header)
    this.link = data.toString('ascii')
  }
}

export class BluetoothPairedList extends Message {
  constructor(header, data) {
    super(header)
    this.data = data.toString('ascii')
  }
}

export const PhoneType = {
  AndroidMirror: 1,
  CarPlay: 3,
  iPhoneMirror: 4,
  AndroidAuto: 5,
  HiCar: 6,
}

export class Plugged extends Message {
  constructor(header, data) {
    super(header)
    const wifiAvail = Buffer.byteLength(data) === 8
    if (wifiAvail) {
      this.phoneType = data.readUInt32LE(0)
      this.wifi = data.readUInt32LE(4)
      console.debug(
        'wifi avail, phone type: ',
        Object.keys(PhoneType).find(k => PhoneType[k] === this.phoneType),
        ' wifi: ',
        this.wifi,
      )
    } else {
      this.phoneType = data.readUInt32LE(0)
      console.debug('no wifi avail, phone type: ', Object.keys(PhoneType).find(k => PhoneType[k] === this.phoneType))
    }
  }
}

export class Unplugged extends Message {
  constructor(header) {
    super(header)
  }
}

export const decodeTypeMap = {
  1: {
    frequency: 44100,
    channel: 2,
    bitDepth: 16,
    format: "S16LE",
    mimeType: "audio/L16; rate=44100; channels=2"
  },
  2: {
    frequency: 44100,
    channel: 2,
    bitDepth: 16,
    format: "S16LE",
    mimeType: "audio/L16; rate=44100; channels=2"
  },
  3: {
    frequency: 8000,
    channel: 1,
    bitDepth: 16,
    format: "S16LE",
    mimeType:  "audio/L16; rate=8000; channels=1"
  },
  4: {
    frequency: 48000,
    channel: 2,
    bitDepth: 16,
    format: "S16LE",
    mimeType:  "audio/L16; rate=48000; channels=2"
  },
  5: {
    frequency: 16000,
    channel: 1,
    bitDepth: 16,
    format: "S16LE",
    mimeType:  "audio/L16; rate=16000; channels=1"
  },
  6: {
    frequency: 24000,
    channel: 1,
    bitDepth: 16,
    format: "S16LE",
    mimeType:  "audio/L16; rate=24000; channels=1"
  },
  7: {
    frequency: 16000,
    channel: 2,
    bitDepth: 16,
    format: "S16LE",
    mimeType:  "audio/L16; rate=16000; channels=2"
  },
}

export class AudioData extends Message {
  constructor(header, data) {
    super(header)
    this.decodeType = data.readUInt32LE(0)
    this.volume = data.readFloatLE(4)
    this.audioType = data.readUInt32LE(8)
    const amount = data.length - 12
    if (amount === 1) {
      this.command = data.readInt8(12)
    } else if (amount === 4) {
      this.volumeDuration = data.readFloatLE(12)
    } else {
      this.data = new Int16Array(data.buffer, 12)
    }
  }
}

export class VideoData extends Message {
  constructor(header, data) {
    super(header)
    this.width = data.readUInt32LE(0)
    this.height = data.readUInt32LE(4)
    this.flags = data.readUInt32LE(8)
    this.length = data.readUInt32LE(12)
    this.unknown = data.readUInt32LE(16)
    this.data = data.subarray(20)
  }
}

const MediaType = {
  Data: 1,
  AlbumCover: 3,
}

export class MediaData extends Message {
  constructor(header, data) {
    super(header)
    const type = data.readUInt32LE(0)
    if (type === MediaType.AlbumCover) {
      const imageData = data.subarray(4)
      this.payload = {
        type,
        base64Image: imageData.toString('base64'),
      }
    } else if (type === MediaType.Data) {
      const mediaData = data.subarray(4, data.length - 1)
      this.payload = {
        type,
        media: JSON.parse(mediaData.toString('utf8')),
      }
    } else {
      console.info(`Unexpected media type: ${type}`)
    }
  }
}

export class Opened extends Message {
  constructor(header, data) {
    super(header)
    this.width = data.readUInt32LE(0)
    this.height = data.readUInt32LE(4)
    this.fps = data.readUInt32LE(8)
    this.format = data.readUInt32LE(12)
    this.packetMax = data.readUInt32LE(16)
    this.iBox = data.readUInt32LE(20)
    this.phoneMode = data.readUInt32LE(24)
  }
}

export class BoxInfo extends Message {
  constructor(header, data) {
    super(header)
    this.settings = JSON.parse(data.toString())
  }
}

export class Phase extends Message {
  constructor(header, data) {
    super(header)
    this.phase = data.readUInt32LE(0)
  }
}
